### PR TITLE
Linker to NativeAOT sync

### PR DESCRIPTION
**Pull request by vitek-karas**
Syncs latest linker changes to NativeAOT
* Improvements in the compiler generated code space
* Fixes a source build problem
* Sync the tests which are already enabled in NativeAOT

### CODE REVIEW

File: src/coreclr/tools/aot/ILLink.Shared/TrimAnalysis/HandleCallAction.cs [hunk 1]
```diff
@@ -783,7 +783,7 @@ GenericParameterValue genericParam
 							// Nothing to do - this throws at runtime
 							AddReturnValue (MultiValueLattice.Top);
 						} else if (typeNameValue is ValueWithDynamicallyAccessedMembers valueWithDynamicallyAccessedMembers && valueWithDynamicallyAccessedMembers.DynamicallyAccessedMemberTypes != 0) {
-							// Propagate the annotation from the type name to the return value. Annotation on a string value will be fulfilled whenever a value is assigned to the string with annotation.
+							// Propagate the annotation from the type name to the return value. Annotation on a string value will be fullfilled whenever a value is assigned to the string with annotation.
```
Comment (am11) [inline src/coreclr/tools/aot/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:786]: `fulfilled` is correct. Could you please fix it in source build.

File: src/coreclr/tools/aot/ILLink.Shared/TrimAnalysis/HandleCallAction.cs [hunk 2]
```diff
@@ -1187,7 +1178,7 @@ GenericParameterValue genericParam
 						if (!methodReturnValueWithMemberTypes.DynamicallyAccessedMemberTypes.HasFlag (annotatedMethodReturnValue.DynamicallyAccessedMemberTypes))
 							throw new InvalidOperationException ($"Internal linker error: in {GetContainingSymbolDisplayName ()} processing call to {calledMethod.GetDisplayName ()} returned value which is not correctly annotated with the expected dynamic member access kinds.");
 					} else if (uniqueValue is SystemTypeValue) {
-						// SystemTypeValue can fulfill any requirement, so it's always valid
+						// SystemTypeValue can fullfill any requirement, so it's always valid
```
Comment (am11) [inline src/coreclr/tools/aot/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:1181]: Same here.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs [hunk 3]
```diff
@@ -6,6 +6,7 @@
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using System.Linq;
+using System.Reflection.Emit;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs:9]: Nit: Is this used?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs [hunk 4]
```diff
@@ -111,58 +112,87 @@ void ProcessMethod(MethodDesc method)
...
                         {
                             ILOpcode opcode = reader.ReadILOpcode();
-                            MethodDesc? lambdaOrLocalFunction;
                             switch (opcode)
                             {
                                 case ILOpcode.ldftn:
                                 case ILOpcode.ldtoken:
                                 case ILOpcode.call:
                                 case ILOpcode.callvirt:
+                                case ILOpcode.initobj:
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs:121]: initobj is not like the others - it takes a type token, not method token. It doesn't look right given what we do below.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs [hunk 5]
```diff
@@ -352,27 +377,68 @@ void MapGeneratedTypeTypeParameters(MetadataType generatedType)
...
                 {
                     ILReader reader = new ILReader(body.GetILBytes());
                     while (reader.HasNext)
                     {
                         ILOpcode opcode = reader.ReadILOpcode();
+                        bool handled = false;
+                        MethodDesc? methodOperand = null;
                         switch (opcode)
                         {
                             case ILOpcode.initobj:
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs:390]: Same here - initobj is a type.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs [hunk 6]
```diff
@@ -352,27 +377,68 @@ void MapGeneratedTypeTypeParameters(MetadataType generatedType)
...
                             default:
                                 reader.Skip(opcode);
                                 break;
                         }
+
+                        // Also look for type substitutions into generic methods
+                        // (such as AsyncTaskMethodBuilder::Start<TStateMachine>).
+                        if (!handled && methodOperand is not null)
+                        {
+                            if (methodOperand != methodOperand.GetTypicalMethodDefinition())
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs:430]: Nit: we don't need to enter the foreach for methods on generic types, only generic methods.

```suggestion
                            if (methodOperand != methodOperand.GetMethodDefinition())
```

**Review by MichalStrehovsky — APPROVED** [review]
Thank you!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs (hunk 4 above)
Comment (vitek-karas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs:121]: I made this mistake in a previous change here and now just duplicated it. Removed initobj.

**Comment by vitek-karas** [comment]
The test failure are all known. Most of them were "Fixed" by this PR https://github.com/dotnet/runtime/pull/73476.

**Pull request merged**